File: 1706. Where Will the Ball Fall/dfs.cpp
class Solution {
public:
    vector<int> findBall(vector<vector<int>>& grid) {
        vector<int> res;
        int row = grid.size(), col = grid[0].size();
        for (int i = 0; i < col; i++)
            res.push_back(dfs(grid, 0, i, row, col));
        return res;
    }

    int dfs(vector<vector<int>>& grid, int r, int c, int row, int col) {
        if (r == row)
            return c;
        else if (grid[r][c] == 1 && c+1 < col && grid[r][c+1] == 1)
            return dfs(grid, r+1, c+1, row, col);
        else if (grid[r][c] == -1 && c-1 >= 0 && grid[r][c-1] == -1)
            return dfs(grid, r+1, c-1, row, col);
        else
            return -1;
    }
};

File: 1706. Where Will the Ball Fall/dfs.py
class Solution:
    def findBall(self, grid: List[List[int]]) -> List[int]:
        row, col, res = len(grid), len(grid[0]), []
        for c in range(col):
            res.append(self.dfs(grid, 0, c, row, col))
        return res
    
    def dfs(self, grid, r, c, row, col):
        if r == row:
            return c
        elif grid[r][c] == 1 and c+1 < col and grid[r][c+1] == 1:
            return self.dfs(grid, r+1, c+1, row, col)
        elif grid[r][c] == -1 and c-1 >= 0 and grid[r][c-1] == -1:
            return self.dfs(grid, r+1, c-1, row, col)
        else:
            return -1